### ISSUE TITLE
Question about regular's regular expressions

### ISSUE THREAD

**Issue by matsc-at-sics-se** [MiniZinc/libminizinc#438]
The regular expression syntax for:

`predicate regular(array [int] of var int: x, string: r)
`

mentions Union, Groups, Concatenation, Wildcards etc., but **not** Intersection.
Intersection is essential if you want to apply two regular expressions to the same sequence in one go, e.g. for domain consistency. Am I missing something? If I am not, consider this a feature (intersection) request.

**Comment by Dekker1** [comment]
Hi Mats,

The idea behind the string version of `regular` is currently that it closely follows the syntax of regular expressions as used in many other programming languages like POSIX shell and perl. To my knowledge there is no real syntax to express the intersection of two regular expression in the common syntax. The closest I can find is the use of two lookahead expressions: https://stackoverflow.com/questions/24102484/can-regex-match-intersection-between-two-regular-expressions, but as noted in the answers this could make the expression non-regular.

Unless there is some part of regular expression syntax that I'm missing, I assume that the general idea for these regular expressions (in other languages) would be that the programmer would combine the patterns manually.

There is indeed something to say for combining NFAs/DFAs when applying them on the same array of variables, but I'm not sure in what way it should be added (and if it is an important feature to add):
-  I'm not a big fan of introducing new syntax for regular expressions (even though we cheating a little bit with spacing in the syntax to begin with).
- Looking at some examples on the internet it generally seems like it should generally be possible (and not impractical) to manually craft the intersection of two regular expressions.
- I'm unsure if this situation occurs often while modelling a problem. When modelling a problem it seem odd to me to describe the sequence of values that an array has to take using multiple expressions.

Note that the technique to craft intersections in the compiler should already exist, but I'm just not convinced what complications would be worth this extra functionality. One idea that came to mind while typing this would be to introduce:
```
predicate regular(array [int] of var int: x, array[int] of string: r)
```
where all regular expressions in the array `r` would be applied to the array `x`, but again I'm not sure if it wouldn't be confusing.

Maybe other modellers or members of the MiniZinc team could chime in on if this is actually a feature that is missing and if so, how we can introduce it without complicating the library or the regular expression syntax.

**Comment by zayenz** [comment]
As a reference, I once implemented intersection for regular constraints in Gecode. At that point I could not find any cases where it was actually useful, in large part due to the quadratic blowup that occurs.

**Comment by matsc-at-sics-se** [comment]
Re. usefulness, I thought it was common practice, or at least folklore, to use the intersection of regexps (the product of DFAs) to get a DC constraint over a sequence. See e.g, ["On matrices, automata, and double counting in constraint programming"](https://link.springer.com/content/pdf/10.1007/s10601-012-9134-y.pdf).

Re. blowup, yes, the blowup is quadratic in the worst case, but often much less than that, thanks to DFA minimization. That's the price you have to pay for domain consistency.

Re. syntax, you just have to pick some infix operator. I would pick "&" to make it symmetric to "|".

Re:
`predicate regular(array [int] of var int: x, array[int] of string: r
`
I think that's a bad idea. IMHO it's better to make the regexp language truly compositional. On that note, I'm also missing Complement and/or Difference; one of the two suffices.

Re: crafting the intersection of two regexps, I don't understand your point. Behind the scenes, I suppose the regexp is transformed to a DFA. To support intersection, you need to be able to compute the product of two DFAs. That's a textbook algorithm, and so is DFA minimization, which I assume that you are using already.

**Comment by zayenz** [comment]
I agree that using intersection is a known:ish technique for those that care (when I experimented with it, it was because it was the obvious thing to try), but intersection of finite automata is also one of those things that are typically shown to be doable but not considered very practically relevant due to the blowup. 

In most of the cases where I've used regular (either with a finite automaton specification or a regexp), the constructions have been easy to do for each individual constraint, and combining the constraints (while theoretically more powerful) is often not that interesting since there is not a lot of extra pruning. This of course entirely depends on the problem at hand.

If the feature was generally available in MiniZinc, then it would perhaps invite more experimentation and use, but I do wonder about how often that would be. I agree that syntax for the construction is probably a good idea compared with the array of expressions, and that `&` is a good symbol. Negation would be similarly useful, since it is sometimes much easier to specify negative patterns.

**Comment by matsc-at-sics-se** [comment]
If you don't want to extend the regexp syntax, consider adding regexp combinator functions:
```
function string: regexp_intersection(string: r1, string: r2);

function string: regexp_difference(string: r1, string: r2);
```
and optionally the following, which can be done by gluing strings together:
```
function string: regexp_union(string: r1, string: r2);

function string: regexp_concatenation(string: r1, string: r2);

function string: regexp_closure(string: r1);
```
